### PR TITLE
Spanner

### PR DESCRIPTION
**Pull request by MarkWolters**
gcp cloud spanner implementation

### CODE REVIEW

File: mvn-defaults/pom.xml [hunk 1]
```diff
@@ -415,6 +415,12 @@
             <version>1.13.2</version>
           </dependency>
 
+            <dependency>
+                <groupId>com.google.cloud</groupId>
+                <artifactId>google-cloud-spanner</artifactId>
+                <version>6.74.0</version>
```
Comment (msmygit) [inline mvn-defaults/pom.xml:421]:
```suggestion
                <version>6.76.0</version>
```

Let's maybe use [6.76.0 here, which is the latest](https://mvnrepository.com/artifact/com.google.cloud/google-cloud-spanner)?

File: nb-adapters/adapter-gcp-spanner/src/main/java/io/nosqlbench/adapter/gcpspanner/GCPSpannerAdapterUtils.java [hunk 2]
```diff
@@ -0,0 +1,34 @@
+/*
+ * Copyright (c) 2024 nosqlbench
```
Comment (msmygit) [inline nb-adapters/adapter-gcp-spanner/src/main/java/io/nosqlbench/adapter/gcpspanner/GCPSpannerAdapterUtils.java:2]:
```suggestion
 * Copyright (c) 2020-2024 nosqlbench
```

File: nb-adapters/adapter-gcp-spanner/src/main/java/io/nosqlbench/adapter/gcpspanner/GCPSpannerDriverAdapter.java [hunk 3]
```diff
@@ -0,0 +1,55 @@
+/*
+ * Copyright (c) 2024 nosqlbench
```
Comment (msmygit) [inline nb-adapters/adapter-gcp-spanner/src/main/java/io/nosqlbench/adapter/gcpspanner/GCPSpannerDriverAdapter.java:2]:
```suggestion
 * Copyright (c) 2020-2024 nosqlbench
```

File: nb-adapters/adapter-gcp-spanner/src/main/java/io/nosqlbench/adapter/gcpspanner/GCPSpannerSpace.java [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+/**
+ * @see <a href="https://cloud.google.com/spanner/docs/find-approximate-nearest-neighbors">ANN Docs</a>
+ * @see <a href="https://github.com/googleapis/java-spanner">Spanner Java Client</a>
+ * @see <a href="https://cloud.google.com/spanner/docs/getting-started/java">Getting started in Java</a>
+ * @see <a href="https://cloud.google.com/docs/authentication#getting_credentials_for_server-centric_flow">Authentication methods at Google</a>
+ * @see <a href="https://cloud.google.com/java/docs/reference/google-cloud-spanner/latest/overview">Library Reference Doc</a>
+ * @see <a href="https://cloud.google.com/spanner/docs/reference/standard-sql/dml-syntax">DML Syntax</a>
+ * @see <a href="https://cloud.google.com/spanner/docs/reference/rpc">spanner rpc api calls</a>
+ * @see <a href="https://cloud.google.com/spanner/docs/reference/standard-sql/data-definition-language#vector_index_statements">SQL functionality related to vector indices</a>
+ * @see <a href=""></a>
```
Comment (msmygit) [inline nb-adapters/adapter-gcp-spanner/src/main/java/io/nosqlbench/adapter/gcpspanner/GCPSpannerSpace.java:36]: sorry I left them on here; I guess we remove these empty ones.

File: nb-adapters/adapter-gcp-spanner/src/main/java/io/nosqlbench/adapter/gcpspanner/GCPSpannerSpace.java [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+
+    public String getDatabaseIdString() {
+        return cfg.get("database_id");
+    }
+
+    private void createSpanner() {
+        if (
+            cfg.getOptional("database_id").isEmpty() ||
+            cfg.getOptional("project_id").isEmpty() ||
+            cfg.getOptional("instance_id").isEmpty()) {
```
Comment (msmygit) [inline nb-adapters/adapter-gcp-spanner/src/main/java/io/nosqlbench/adapter/gcpspanner/GCPSpannerSpace.java:92]:
```suggestion
            cfg.getOptional("instance_id").isEmpty() || cfg.getOptional("service_account_file").isEmpty()) {
```

File: nb-adapters/adapter-gcp-spanner/src/main/java/io/nosqlbench/adapter/gcpspanner/opdispensers/GCPSpannerBaseOpDispenser.java [hunk 6]
```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright (c) 2024 nosqlbench
```
Comment (msmygit) [inline nb-adapters/adapter-gcp-spanner/src/main/java/io/nosqlbench/adapter/gcpspanner/opdispensers/GCPSpannerBaseOpDispenser.java:2]:
```suggestion
 * Copyright (c) 2020-2024 nosqlbench
```

File: nb-adapters/adapter-gcp-spanner/src/main/java/io/nosqlbench/adapter/gcpspanner/opdispensers/GCPSpannerExecuteDmlOpDispenser.java [hunk 7]
```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright (c) 2024 nosqlbench
```
Comment (msmygit) [inline nb-adapters/adapter-gcp-spanner/src/main/java/io/nosqlbench/adapter/gcpspanner/opdispensers/GCPSpannerExecuteDmlOpDispenser.java:2]:
```suggestion
 * Copyright (c) 2020-2024 nosqlbench
```

File: nb-adapters/adapter-gcp-spanner/src/main/java/io/nosqlbench/adapter/gcpspanner/opdispensers/GCPSpannerInsertOpDispenser.java [hunk 8]
```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright (c) 2024 nosqlbench
```
Comment (msmygit) [inline nb-adapters/adapter-gcp-spanner/src/main/java/io/nosqlbench/adapter/gcpspanner/opdispensers/GCPSpannerInsertOpDispenser.java:2]:
```suggestion
 * Copyright (c) 2020-2024 nosqlbench
```

File: nb-adapters/adapter-gcp-spanner/src/main/java/io/nosqlbench/adapter/gcpspanner/opdispensers/GCPSpannerUpdateDatabaseDdlOpDispenser.java [hunk 9]
```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright (c) 2024 nosqlbench
```
Comment (msmygit) [inline nb-adapters/adapter-gcp-spanner/src/main/java/io/nosqlbench/adapter/gcpspanner/opdispensers/GCPSpannerUpdateDatabaseDdlOpDispenser.java:2]:
```suggestion
 * Copyright (c) 2020-2024 nosqlbench
```

File: nb-adapters/adapter-gcp-spanner/src/main/java/io/nosqlbench/adapter/gcpspanner/ops/GCPSpannerExecuteDmlOp.java [hunk 10]
```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright (c) 2024 nosqlbench
```
Comment (msmygit) [inline nb-adapters/adapter-gcp-spanner/src/main/java/io/nosqlbench/adapter/gcpspanner/ops/GCPSpannerExecuteDmlOp.java:2]:
```suggestion
 * Copyright (c) 2020-2024 nosqlbench
```

File: nb-adapters/adapter-gcp-spanner/src/main/java/io/nosqlbench/adapter/gcpspanner/ops/GCPSpannerInsertOp.java [hunk 11]
```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright (c) 2024 nosqlbench
```
Comment (msmygit) [inline nb-adapters/adapter-gcp-spanner/src/main/java/io/nosqlbench/adapter/gcpspanner/ops/GCPSpannerInsertOp.java:2]:
```suggestion
 * Copyright (c) 2020-2024 nosqlbench
```

File: nb-adapters/adapter-gcp-spanner/src/main/java/io/nosqlbench/adapter/gcpspanner/ops/GCPSpannerUpdateDatabaseDdlOp.java [hunk 12]
```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright (c) 2024 nosqlbench
```
Comment (msmygit) [inline nb-adapters/adapter-gcp-spanner/src/main/java/io/nosqlbench/adapter/gcpspanner/ops/GCPSpannerUpdateDatabaseDdlOp.java:2]:
```suggestion
 * Copyright (c) 2020-2024 nosqlbench
```

**Review by msmygit — DISMISSED** [review]
LGTM 👍🏼

**Review by msmygit — DISMISSED** [review]
LGTM 👍🏼

File: nb-adapters/adapter-gcp-spanner/src/main/resources/gcp-spanner.md [hunk 13]
```diff
@@ -0,0 +1,30 @@
+# Google Spanner driver adapter
```
Comment (jshook) [inline nb-adapters/adapter-gcp-spanner/src/main/resources/gcp-spanner.md:1]: The name of this file and the adapter name should match exactly. Can we simplify them both to just `spanner` or at least `gcp_spanner` in both cases?

**Review by jshook — CHANGES_REQUESTED** [review]
Generally looks good, only one naming change request.

File: nb-adapters/adapter-gcp-spanner/src/main/java/io/nosqlbench/adapter/gcpspanner/GCPSpannerSpace.java (hunk 4 above)
Comment (MarkWolters) [reply nb-adapters/adapter-gcp-spanner/src/main/java/io/nosqlbench/adapter/gcpspanner/GCPSpannerSpace.java]: removed in latest push

File: nb-adapters/adapter-gcp-spanner/src/main/java/io/nosqlbench/adapter/gcpspanner/GCPSpannerSpace.java (hunk 5 above)
Comment (MarkWolters) [reply nb-adapters/adapter-gcp-spanner/src/main/java/io/nosqlbench/adapter/gcpspanner/GCPSpannerSpace.java]: done

File: nb-adapters/adapter-gcp-spanner/src/main/resources/gcp-spanner.md (hunk 13 above)
Comment (MarkWolters) [reply nb-adapters/adapter-gcp-spanner/src/main/resources/gcp-spanner.md]: renamed in both places to simply 'spanner'

**Review by msmygit — DISMISSED** [review]

**Review by msmygit — APPROVED** [review]

**Review by jshook — APPROVED** [review]
LGTM!

**Pull request merged**